File: prioritarr/webhooks/sonarr_grab.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from datetime import datetime, timezone

from prioritarr.database import Database

CLIENT_MAP: dict[str, str] = {
    "qbittorrent": "qbit",
    "sabnzbd": "sab",
    "nzbget": "sab",
}


@dataclass
class OnGrabEvent:
    series_id: int
    series_title: str
    tvdb_id: int
    episode_ids: list[int]
    download_client: str  # normalized via CLIENT_MAP
    download_id: str
    air_date: str | None


def parse_ongrab_payload(data: dict) -> OnGrabEvent:
    """Extract fields from a Sonarr OnGrab webhook payload.

    Normalises the download client name via CLIENT_MAP (falls back to
    the raw value lowercased if unknown).
    """
    series = data.get("series", {})
    episodes = data.get("episodes", [])

    raw_client = data.get("downloadClient", "")
    normalized_client = CLIENT_MAP.get(raw_client.lower(), raw_client.lower())

    episode_ids = [ep["id"] for ep in episodes if "id" in ep]

    # Use the airDateUtc of the first episode (if present)
    air_date: str | None = None
    if episodes:
        air_date = episodes[0].get("airDateUtc")

    return OnGrabEvent(
        series_id=series.get("id", 0),
        series_title=series.get("title", ""),
        tvdb_id=series.get("tvdbId", 0),
        episode_ids=episode_ids,
        download_client=normalized_client,
        download_id=data.get("downloadId", ""),
        air_date=air_date,
    )


def _event_key(event: OnGrabEvent) -> str:
    """Compute a SHA-1 deduplication key from the event's identifying fields."""
    raw = f"Grab:{event.series_id}:{sorted(event.episode_ids)}:{event.download_id}"
    return hashlib.sha1(raw.encode()).hexdigest()


def handle_ongrab(
    event: OnGrabEvent,
    db: Database,
    priority: int,
    dry_run: bool = False,
) -> bool:
    """Process an OnGrab event.

    Deduplicates via db.try_insert_dedupe, inserts a managed_download row,
    and appends an audit entry.

    Returns True if the event was processed, False if it was a duplicate.
    """
    now = datetime.now(timezone.utc).isoformat()
    key = _event_key(event)

    if not db.try_insert_dedupe(key, now):
        return False

    if not dry_run:
        db.upsert_managed_download(
            client=event.download_client,
            client_id=event.download_id,
            series_id=event.series_id,
            episode_ids=event.episode_ids,
            initial_priority=priority,
            current_priority=priority,
            paused_by_us=False,
            first_seen_at=now,
            last_reconciled_at=now,
        )

    db.append_audit(
        action="ongrab",
        series_id=event.series_id,
        client=event.download_client,
        client_id=event.download_id,
        details={
            "series_title": event.series_title,
            "episode_ids": event.episode_ids,
            "priority": priority,
            "dry_run": dry_run,
        },
    )

    return True

File: prioritarr/backend/src/main/kotlin/org/yoshiz/app/prioritarr/backend/webhooks/SonarrGrab.kt
package org.yoshiz.app.prioritarr.backend.webhooks

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put
import kotlinx.serialization.json.add
import org.yoshiz.app.prioritarr.backend.database.Database
import java.security.MessageDigest

/** Mirrors CLIENT_MAP in prioritarr/webhooks/sonarr_grab.py. */
val CLIENT_MAP = mapOf(
    "qbittorrent" to "qbit",
    "sabnzbd" to "sab",
    "nzbget" to "sab",
)

data class OnGrabEvent(
    val seriesId: Long,
    val seriesTitle: String,
    val tvdbId: Long,
    val episodeIds: List<Long>,
    val downloadClient: String,
    val downloadId: String,
    val airDate: String?,
)

fun parseOnGrabPayload(data: JsonObject): OnGrabEvent {
    val series = (data["series"] as? JsonObject) ?: JsonObject(emptyMap())
    val episodes = (data["episodes"]?.jsonArray) ?: kotlinx.serialization.json.JsonArray(emptyList())

    val rawClient = data["downloadClient"]?.jsonPrimitive?.contentOrNull.orEmpty().lowercase()
    val normalizedClient = CLIENT_MAP[rawClient] ?: rawClient

    val episodeIds = episodes.mapNotNull {
        it.jsonObject["id"]?.jsonPrimitive?.longOrNull
    }

    val airDate = episodes.firstOrNull()?.jsonObject?.get("airDateUtc")?.jsonPrimitive?.contentOrNull

    return OnGrabEvent(
        seriesId = series["id"]?.jsonPrimitive?.longOrNull ?: 0L,
        seriesTitle = series["title"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        tvdbId = series["tvdbId"]?.jsonPrimitive?.longOrNull ?: 0L,
        episodeIds = episodeIds,
        downloadClient = normalizedClient,
        downloadId = data["downloadId"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        airDate = airDate,
    )
}

fun eventKey(event: OnGrabEvent): String {
    val raw = "Grab:${event.seriesId}:${event.episodeIds.sorted()}:${event.downloadId}"
    val sha = MessageDigest.getInstance("SHA-1").digest(raw.toByteArray())
    return sha.joinToString("") { "%02x".format(it) }
}

/**
 * Deduplicate, write managed_download row (unless dry-run), append audit.
 * Returns true if processed, false if a duplicate.
 */
fun handleOnGrab(
    event: OnGrabEvent,
    db: Database,
    priority: Int,
    dryRun: Boolean = false,
): Boolean {
    val now = Database.nowIsoOffset()
    val key = eventKey(event)

    if (!db.tryInsertDedupe(key, now)) return false

    if (!dryRun) {
        db.upsertManagedDownload(
            client = event.downloadClient,
            clientId = event.downloadId,
            seriesId = event.seriesId,
            episodeIds = event.episodeIds,
            initialPriority = priority.toLong(),
            currentPriority = priority.toLong(),
            pausedByUs = false,
            firstSeenAt = now,
            lastReconciledAt = now,
        )
    }

    db.appendAudit(
        action = "ongrab",
        seriesId = event.seriesId,
        client = event.downloadClient,
        clientId = event.downloadId,
        details = buildJsonObject {
            put("series_title", event.seriesTitle)
            put("episode_ids", kotlinx.serialization.json.buildJsonArray {
                event.episodeIds.forEach { add(it) }
            })
            put("priority", priority)
            put("dry_run", dryRun)
        },
    )
    return true
}
